libcore/ListModel.vala

```diff
@@ -0,0 +1,717 @@
+        directory_reverse_map.unset (subdir);
+        file_entry.loaded = false;
+
+        /* Remove all children */
+        while (file_entry.files.get_length () > 0) {
+            var child_seq = file_entry.files.get_begin_iter ();
+            unowned FileEntry child_file_entry = child_seq.get ();
+            if (child_file_entry.file == null) {
+                /* Don't delete the dummy node */
+                break;
+            } else {
+                Gtk.TreeIter child_iter;
+                sequenceiter_to_treeiter (child_seq, out child_iter);
+                remove (child_iter);
+            }
+        }
+
+        subdirectory_unloaded (subdir);
+        return true;
+    }
+
+    public bool add_file (GOF.File file, GOF.Directory.Async dir) {
+        var parent_seq = directory_reverse_map.get (dir);
+        GLib.SequenceIter<FileEntry?> seq = null;
+        if (parent_seq != null) {
+            unowned FileEntry file_entry = parent_seq.get ();
+            seq = file_entry.reverse_map.get (file);
+        } else {
+            seq = top_reverse_map.get (file);
+        }
+
+        if (seq != null) {
+            return false;
+        }
+
+        var file_entry = FileEntry ();
+        file_entry.file = file;
+
+        unowned GLib.Sequence<FileEntry?> current_files = files;
+        var parent_hash = top_reverse_map;
+        var replaced_dummy = false;
+
+        if (parent_seq != null) {
+            unowned FileEntry parent_entry = parent_seq.get ();
+            file_entry.parent = &parent_entry;
+            /* At this point we set loaded. Either we saw
+             * "done" and ignored it waiting for this, or we do this
+             * earlier, but then we replace the dummy row anyway,
+             * so it doesn't matter */
+            parent_entry.loaded = true;
+            parent_hash = parent_entry.reverse_map;
+            current_files = parent_entry.files;
+            /* maybe the dummy row */
+            if (current_files.get_length () == 1) {
+                var dummy_seq = current_files.get_iter_at_pos (0);
+                unowned FileEntry dummy_entry = dummy_seq.get ();
+                /* it is the dummy row  - replace it */
+                if (dummy_entry.file == null) {
+                    dummy_seq.remove ();
+                    replaced_dummy = true;
+                }
+            }
+        }
+
+        file_entry.seq = current_files.insert_sorted (file_entry, file_entry_compare_func);
+        parent_hash.set (file, file_entry.seq);
+
+        var iter = Gtk.TreeIter ();
+        iter.stamp = stamp;
+        iter.user_data = file_entry.seq;
+
+        var path = get_path (iter);
+        if (replaced_dummy) {
+            row_changed (path, iter);
+        } else {
+            row_inserted (path, iter);
+        }
+
+        if (file.is_folder ()) {
+            var file_entry_files = new GLib.Sequence<FileEntry?> ();
+            file_entry.files = file_entry_files;
```

Comment:
This does not work.  The new `Sequence<FileEntry?>` gets freed as soon as the clause exits.  This is why no expanders appear on folders in ListView. However the obvious port of the C `file_entry.files = new Sequence<FileEntry?>()` gives an error both when file_entry.files is owned (duplicating a sequence) or unowned (assigning owned to unowned).  Not sure what the solution is to this.  Is Vala behaving correctly here?  May need to use a different structure for FileEntry and/or Sequence